File: ios/TimberCruisingApp/ViewModels/CrashRecoveryService.swift
// Phase 7 — crash-recovery resume prompt.
//
// On app launch we look for any plots that were *opened* (closedAt ==
// nil) in the last 24 hours. If found, the home screen shows a
// three-option prompt:
//
//   • Yes, resume  — reopen the plot's tally screen where the cruiser left off
//   • View         — show plot #, tree count, last-edited time before deciding
//   • No, discard  — keep the plot in place but dismiss the prompt (nothing
//                    is deleted; cruiser can still open it later from the
//                    project dashboard)
//
// The 24-hour threshold strikes a balance: a week-old half-finished plot
// is almost certainly abandoned, but today's / yesterday's work is worth
// surfacing. The cutoff is configurable so unit tests can inject a
// deterministic clock.

import Foundation
import Models
import Persistence

public struct ResumeCandidate: Identifiable, Sendable {
    public var id: UUID { plot.id }
    public let plot: Plot
    public let projectName: String
    public let liveTreeCount: Int
    public let lastEditedAt: Date

    public var summary: String {
        "Plot \(plot.plotNumber) • \(liveTreeCount) trees • " +
        RelativeDateTimeFormatter().localizedString(
            for: lastEditedAt, relativeTo: Date())
    }
}

public enum CrashRecoveryService {

    /// Scan every project for open plots younger than `maxAge` whose last
    /// activity (max of `startedAt` and any contained tree's `updatedAt`)
    /// falls inside the window. Returns the candidates sorted most-recent
    /// first.
    public static func openPlotsWithinLast(
        _ maxAge: TimeInterval,
        projectRepo: any ProjectRepository,
        plotRepo: any PlotRepository,
        treeRepo: any TreeRepository,
        now: Date = Date()
    ) throws -> [ResumeCandidate] {
        let cutoff = now.addingTimeInterval(-maxAge)
        var candidates: [ResumeCandidate] = []
        for project in try projectRepo.list() {
            let plots = try plotRepo.listByProject(project.id)
            for plot in plots where plot.closedAt == nil {
                let trees = try treeRepo.listByPlot(plot.id, includeDeleted: false)
                let last = (trees.map { $0.updatedAt } + [plot.startedAt]).max()
                    ?? plot.startedAt
                guard last >= cutoff else { continue }
                candidates.append(ResumeCandidate(
                    plot: plot,
                    projectName: project.name,
                    liveTreeCount: trees.count,
                    lastEditedAt: last))
            }
        }
        return candidates.sorted { $0.lastEditedAt > $1.lastEditedAt }
    }
}

File: android/app/src/main/java/com/hcjeong/forestix/data/cruise/CrashRecoveryService.kt
// Crash-recovery resume scan — Android counterpart of iOS
// ViewModels/CrashRecoveryService.swift (Phase 7).
//
// On the home screen's first appearance per launch we look for any plots that
// were *opened* (closedAt == null) whose last activity falls inside the last
// 24 hours. If found, MapHomeScreen shows a three-option prompt:
//
//   • Resume  — reopen the plot's active cruise/tally surface where the
//               cruiser left off (sets the current project + active plot and
//               flips the map into cruise mode)
//   • View    — plot #, tree count and last-edited are shown inline so the
//               cruiser can decide before acting (folded into the prompt)
//   • Discard — dismiss the prompt; NOTHING is deleted, the plot stays open
//               and reachable later from the cruise map
//
// The 24-hour threshold: a week-old half-finished plot is almost certainly
// abandoned, but today's / yesterday's work is worth surfacing. `now` and
// `maxAgeMs` are injectable so tests can pin a deterministic clock.

package com.hcjeong.forestix.data.cruise

object CrashRecoveryService {

    /// 24 hours, in milliseconds — the default resume window.
    const val DEFAULT_MAX_AGE_MS = 24L * 60 * 60 * 1000

    /// Process-wide "already checked this launch" guard. The home screen may
    /// recompose or be re-entered (Settings round-trip, mode toggle) many times
    /// per launch; the prompt must appear AT MOST once, and a dismiss/Discard
    /// must not re-prompt. Set true the first time the scan runs.
    @Volatile
    var checkedThisLaunch: Boolean = false

    /// A resumable open plot plus the display facts the prompt surfaces.
    data class ResumeCandidate(
        val plot: Plot,
        val projectName: String,
        val liveTreeCount: Int,
        /// max(plot.startedAt, newest live tree.updatedAt), epoch millis.
        val lastEditedAt: Long,
    ) {
        val id get() = plot.id
    }

    /// Scan every project for open plots (closedAt == null) whose last activity
    /// (max of `startedAt` and any contained live tree's `updatedAt`) falls
    /// inside `maxAgeMs` of `now`. Returns candidates sorted most-recent first.
    suspend fun openPlotsWithinLast(
        maxAgeMs: Long = DEFAULT_MAX_AGE_MS,
        projectRepo: ProjectRepository,
        plotRepo: PlotRepository,
        treeRepo: TreeRepository,
        now: Long = System.currentTimeMillis(),
    ): List<ResumeCandidate> {
        val cutoff = now - maxAgeMs
        val candidates = mutableListOf<ResumeCandidate>()
        for (project in projectRepo.list()) {
            val plots = plotRepo.listByProject(project.id)
            for (plot in plots) {
                if (plot.closedAt != null) continue
                val trees = treeRepo.listByPlot(plot.id, includeDeleted = false)
                val lastEdited =
                    (trees.map { it.updatedAt } + plot.startedAt).maxOrNull() ?: plot.startedAt
                if (lastEdited < cutoff) continue
                candidates.add(
                    ResumeCandidate(
                        plot = plot,
                        projectName = project.name,
                        liveTreeCount = trees.size,
                        lastEditedAt = lastEdited,
                    )
                )
            }
        }
        return candidates.sortedByDescending { it.lastEditedAt }
    }
}
